File: android/core/src/main/java/tech/robihamanto/heritg/android/core/tree/TreeObstacleRouter.kt
package tech.robihamanto.heritg.android.core.tree

import tech.robihamanto.heritg.android.core.domain.EnglishSemanticFormatter
import tech.robihamanto.heritg.android.core.domain.SemanticFormatter
import tech.robihamanto.heritg.android.core.model.RelationshipKind
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min

data class TreeNonParentRoute(
    val edge: TreeEdgeLayout,
    val segments: List<Segment>,
    val labelPosition: Point?,
    val labelObstacle: TreeObstacle?,
) {
    val id: String get() = edge.id
}

object TreeObstacleRouter {
    internal const val CoordinatePadding = 2.0
    private const val BendPenalty = 24.0
    private const val Epsilon = 0.001

    fun nodeObstacles(layout: TreeLayoutResult, showsRelationshipLabels: Boolean): List<TreeObstacle> =
        layout.nodes.sortedBy { it.id }.flatMap { node ->
            val addSide = TreeVisualMetrics.addControlSide(
                occupiedSides(node, layout.edges),
                if (node.position.x <= 0) TreeNodeSide.LEFT else TreeNodeSide.RIGHT,
            )
            val addCenter = TreeVisualMetrics.addControlPosition(node.position, 1.0, addSide)
            val editCenter = TreeVisualMetrics.adjacentControlPosition(addCenter, 1.0, addSide)
            listOf(
                TreeObstacle(TreeObstacleKind.Avatar(node.id), TreeVisualMetrics.avatarRect(node.position)),
                TreeObstacle(
                    TreeObstacleKind.NodeLabel(node.id),
                    TreeVisualMetrics.nodeLabelRect(
                        node.position,
                        showsRelationshipLabels,
                        node.person.lifeSummary != null,
                    ),
                ),
                TreeObstacle(TreeObstacleKind.AddControl(node.id), TreeVisualMetrics.controlRect(addCenter)),
                TreeObstacle(TreeObstacleKind.EditControl(node.id), TreeVisualMetrics.controlRect(editCenter)),
            )
        }

    fun route(
        edge: TreeEdgeLayout,
        avoiding: List<TreeObstacle>,
        occupiedSegments: List<Segment> = emptyList(),
        semanticFormatter: SemanticFormatter = EnglishSemanticFormatter,
        textMeasurer: TreeTextMeasurer = PortableTreeTextMeasurer,
    ): TreeNonParentRoute {
        val obstacles = sorted(avoiding)
        val endpointIds = setOf(edge.fromPersonId, edge.toPersonId)
        var bestSegments: List<Segment>? = null
        var bestCost = Double.POSITIVE_INFINITY
        terminalPairs(edge).forEach { (penalty, terminals) ->
            val segments = TreeFastRoute.preferredRoute(
                terminals.first,
                terminals.second,
                obstacles,
                endpointIds,
                occupiedSegments,
            ) ?: return@forEach
            val cost = segments.sumOf { it.length } + max(segments.size - 1, 0) * BendPenalty + penalty
            if (cost < bestCost) {
                bestSegments = segments
                bestCost = cost
            }
        }
        val segments = bestSegments.orEmpty()
        val label = relationshipLabelPlacement(
            edge,
            segments,
            obstacles,
            occupiedSegments + segments,
            semanticFormatter,
            textMeasurer,
        )
        return TreeNonParentRoute(edge, segments, label?.first, label?.second)
    }

    fun route(
        segment: Segment,
        avoiding: List<TreeObstacle>,
        endpointPersonIds: Set<String>,
        occupiedSegments: List<Segment> = emptyList(),
    ): List<Segment> = TreeFastRoute.preferredRoute(
        segment.start,
        segment.end,
        sorted(avoiding),
        endpointPersonIds,
        occupiedSegments,
    ).orEmpty()

    fun routeIsClear(
        segments: List<Segment>,
        avoiding: List<TreeObstacle>,
        endpointPersonIds: Set<String> = emptySet(),
    ): Boolean = segments.isNotEmpty() && segments.all { segment ->
        avoiding.all { obstacle -> !hasForbiddenIntersection(segment, obstacle, endpointPersonIds) }
    }

    fun hasForbiddenIntersection(
        segment: Segment,
        obstacle: TreeObstacle,
        endpointPersonIds: Set<String>,
    ): Boolean {
        if (!segmentIntersects(segment, obstacle.rect, TreeVisualMetrics.ConnectorClearance)) return false
        if (!TreeFastRoute.permitsTerminalExit(segment, obstacle, endpointPersonIds)) return true
        return segmentIntersects(segment, obstacle.rect, 0.0)
    }

    fun permitsTerminalContact(point: Point, obstacle: TreeObstacle): Boolean = when (obstacle.kind) {
        is TreeObstacleKind.Avatar -> {
            val touchesVertical = (abs(point.x - obstacle.rect.minX) < Epsilon ||
                abs(point.x - obstacle.rect.maxX) < Epsilon) &&
                point.y >= obstacle.rect.minY - Epsilon && point.y <= obstacle.rect.maxY + Epsilon
            val touchesHorizontal = (abs(point.y - obstacle.rect.minY) < Epsilon ||
                abs(point.y - obstacle.rect.maxY) < Epsilon) &&
                point.x >= obstacle.rect.minX - Epsilon && point.x <= obstacle.rect.maxX + Epsilon
            touchesVertical || touchesHorizontal
        }
        is TreeObstacleKind.NodeLabel -> abs(point.y - obstacle.rect.maxY - 2) < Epsilon &&
            point.x >= obstacle.rect.minX - Epsilon && point.x <= obstacle.rect.maxX + Epsilon
        is TreeObstacleKind.AddControl,
        is TreeObstacleKind.EditControl,
        is TreeObstacleKind.RelationshipLabel,
        -> false
    }

    fun segmentIntersects(segment: Segment, rect: TreeRect, clearance: Double): Boolean {
        val expanded = rect.insetBy(-clearance, -clearance)
        return when (segment.orientation) {
            SegmentOrientation.HORIZONTAL -> {
                if (segment.start.y <= expanded.minY + Epsilon || segment.start.y >= expanded.maxY - Epsilon) false
                else max(min(segment.start.x, segment.end.x), expanded.minX) <
                    min(max(segment.start.x, segment.end.x), expanded.maxX) - Epsilon
            }
            SegmentOrientation.VERTICAL -> {
                if (segment.start.x <= expanded.minX + Epsilon || segment.start.x >= expanded.maxX - Epsilon) false
                else max(min(segment.start.y, segment.end.y), expanded.minY) <
                    min(max(segment.start.y, segment.end.y), expanded.maxY) - Epsilon
            }
            null -> false
        }
    }

    fun hasCollinearOverlap(route: List<Segment>, occupied: List<Segment>): Boolean =
        TreeFastRoute.hasCollinearOverlap(route, occupied)

    internal fun firstRouteCandidate(
        start: Point,
        end: Point,
        obstacles: List<TreeObstacle>,
        isAccepted: (List<Segment>) -> Boolean,
    ): List<Segment>? {
        if (start.x == end.x || start.y == end.y) {
            val candidate = segments(listOf(start, end))
            if (isAccepted(candidate)) return candidate
        }
        val minObstacleY = obstacles.minOfOrNull { it.rect.minY } ?: min(start.y, end.y)
        val maxObstacleY = obstacles.maxOfOrNull { it.rect.maxY } ?: max(start.y, end.y)
        val channelYs = unique(
            buildList {
                add(min(start.y, end.y) - TreeVisualMetrics.AvatarRadius - TreeVisualMetrics.ConnectorClearance)
                add(max(start.y, end.y) + TreeVisualMetrics.AvatarRadius + TreeVisualMetrics.ConnectorClearance)
                add(minObstacleY - TreeVisualMetrics.ConnectorClearance - CoordinatePadding)
                add(maxObstacleY + TreeVisualMetrics.ConnectorClearance + CoordinatePadding)
                obstacles.forEach {
                    add(it.rect.minY - TreeVisualMetrics.ConnectorClearance - CoordinatePadding)
                    add(it.rect.maxY + TreeVisualMetrics.ConnectorClearance + CoordinatePadding)
                }
            },
        ).sortedWith(distanceComparator((start.y + end.y) / 2))
        channelYs.forEach { channelY ->
            val startXs = escapeXCoordinates(start, channelY, obstacles)
            val endXs = escapeXCoordinates(end, channelY, obstacles)
            startXs.forEach { startX ->
                endXs.forEach { endX ->
                    val candidate = segments(
                        listOf(
                            start,
                            Point(startX, start.y),
                            Point(startX, channelY),
                            Point(endX, channelY),
                            Point(endX, end.y),
                            end,
                        ),
                    )
                    if (isAccepted(candidate)) return candidate
                }
            }
        }
        return null
    }

    internal fun segments(rawPoints: List<Point>): List<Segment> {
        val points = rawPoints.fold(mutableListOf<Point>()) { result, point ->
            if (result.lastOrNull() != point) result += point
            result
        }
        if (points.size <= 1) return emptyList()
        var index = 1
        while (index < points.lastIndex) {
            val previous = points[index - 1]
            val current = points[index]
            val next = points[index + 1]
            if ((previous.x == current.x && current.x == next.x) ||
                (previous.y == current.y && current.y == next.y)
            ) points.removeAt(index) else index++
        }
        return points.zipWithNext(::Segment).filter { it.orientation != null }
    }

    private fun terminalPairs(edge: TreeEdgeLayout): List<Pair<Double, Pair<Point, Point>>> {
        val left = if (edge.from.x <= edge.to.x) edge.from else edge.to
        val right = if (edge.from.x <= edge.to.x) edge.to else edge.from
        val radius = TreeVisualMetrics.AvatarRadius
        return listOf(
            0.0 to (Point(left.x + radius, left.y) to Point(right.x - radius, right.y)),
            80.0 to (Point(left.x, left.y - radius) to Point(right.x, right.y - radius)),
            120.0 to (Point(left.x - radius, left.y) to Point(right.x + radius, right.y)),
        )
    }

    private fun escapeXCoordinates(point: Point, toY: Double, obstacles: List<TreeObstacle>): List<Double> {
        val lowerY = min(point.y, toY)
        val upperY = max(point.y, toY)
        val blockers = obstacles.filter { obstacle ->
            point.x > obstacle.rect.minX - TreeVisualMetrics.ConnectorClearance &&
                point.x < obstacle.rect.maxX + TreeVisualMetrics.ConnectorClearance &&
                max(lowerY, obstacle.rect.minY) <= min(upperY, obstacle.rect.maxY)
        }
        val values = buildList {
            add(point.x)
            blockers.forEach {
                add(it.rect.minX - TreeVisualMetrics.ConnectorClearance - CoordinatePadding)
                add(it.rect.maxX + TreeVisualMetrics.ConnectorClearance + CoordinatePadding)
            }
        }
        val result = unique(values).sortedWith(distanceComparator(point.x)).take(9).toMutableList()
        listOfNotNull(
            obstacles.minOfOrNull { it.rect.minX }?.minus(TreeVisualMetrics.ConnectorClearance + CoordinatePadding),
            obstacles.maxOfOrNull { it.rect.maxX }?.plus(TreeVisualMetrics.ConnectorClearance + CoordinatePadding),
        ).forEach { value -> if (result.none { abs(it - value) < Epsilon }) result += value }
        return result
    }

    private fun relationshipLabelPlacement(
        edge: TreeEdgeLayout,
        segments: List<Segment>,
        obstacles: List<TreeObstacle>,
        occupiedSegments: List<Segment>,
        semanticFormatter: SemanticFormatter,
        textMeasurer: TreeTextMeasurer,
    ): Pair<Point, TreeObstacle>? {
        val label = edge.marriageLabel(semanticFormatter) ?: return null
        val horizontal = segments.filter { it.orientation == SegmentOrientation.HORIZONTAL }.sortedWith { first, second ->
            if (first.length != second.length) second.length.compareTo(first.length)
            else compareValuesBy(first, second, { it.start.y }, { it.start.x }, { it.end.x })
        }
        val fractions = (1..<20).map { it / 20.0 }.sortedWith(distanceComparator(0.5))
        horizontal.forEach { segment ->
            fractions.forEach { fraction ->
                val anchor = Point(segment.start.x + (segment.end.x - segment.start.x) * fraction, segment.start.y)
                listOf(-22.0, 22.0, -40.0, 40.0, -58.0, 58.0).forEach { offset ->
                    val position = Point(anchor.x, anchor.y + offset)
                    val obstacle = TreeObstacle(
                        TreeObstacleKind.RelationshipLabel(edge.id),
                        TreeVisualMetrics.relationshipLabelRect(label, position, textMeasurer),
                    )
                    val clearsObstacles = obstacles.all {
                        !it.rect.insetBy(-TreeVisualMetrics.ConnectorClearance, -TreeVisualMetrics.ConnectorClearance)
                            .intersects(obstacle.rect)
                    }
                    val clearsConnectors = occupiedSegments.all {
                        !segmentIntersects(it, obstacle.rect, TreeVisualMetrics.ConnectorClearance)
                    }
                    if (clearsObstacles && clearsConnectors) return position to obstacle
                }
            }
        }
        return null
    }

    private fun occupiedSides(node: TreeNodeLayout, edges: List<TreeEdgeLayout>): Set<TreeNodeSide> = buildSet {
        edges.forEach { edge ->
            val other = when (node.id) {
                edge.fromPersonId -> edge.to
                edge.toPersonId -> edge.from
                else -> return@forEach
            }
            when (edge.kind) {
                RelationshipKind.PARENT -> add(if (other.y < node.position.y) TreeNodeSide.TOP else TreeNodeSide.BOTTOM)
                RelationshipKind.PARTNER, RelationshipKind.SIBLING ->
                    add(if (other.x < node.position.x) TreeNodeSide.LEFT else TreeNodeSide.RIGHT)
            }
        }
    }

    private fun sorted(obstacles: List<TreeObstacle>) = obstacles.sortedWith(
        compareBy<TreeObstacle> { it.kind.stableId }
            .thenBy { it.rect.minY }
            .thenBy { it.rect.minX }
            .thenBy { it.rect.height }
            .thenBy { it.rect.width },
    )

    private fun unique(values: List<Double>) = values.fold(mutableListOf<Double>()) { result, value ->
        if (result.none { abs(it - value) < Epsilon }) result += value
        result
    }

    private fun distanceComparator(origin: Double) = Comparator<Double> { first, second ->
        val firstDistance = abs(first - origin)
        val secondDistance = abs(second - origin)
        if (firstDistance == secondDistance) first.compareTo(second) else firstDistance.compareTo(secondDistance)
    }
}

File: ios/Heritg/TreeObstacleRouter.swift
import CoreGraphics
import Foundation

nonisolated enum TreeObstacleRouter {
    private static let coordinatePadding: CGFloat = 2
    private static let bendPenalty: CGFloat = 24

    static func sortedObstacles(
        _ obstacles: [TreeRoutingGeometry.Obstacle]
    ) -> [TreeRoutingGeometry.Obstacle] {
        obstacles.sorted { left, right in
            let leftKey = "\(left.kind.rawValue):\(left.ownerID)"
            let rightKey = "\(right.kind.rawValue):\(right.ownerID)"
            let comparison = TreeRoutingGeometry.compareText(leftKey, rightKey)
            if comparison != .orderedSame { return comparison == .orderedAscending }
            if left.rect.minY != right.rect.minY { return left.rect.minY < right.rect.minY }
            if left.rect.minX != right.rect.minX { return left.rect.minX < right.rect.minX }
            if left.rect.height != right.rect.height { return left.rect.height < right.rect.height }
            return left.rect.width < right.rect.width
        }
    }

    private static func unique(_ values: [CGFloat]) -> [CGFloat] {
        values.sorted().reduce(into: []) { result, value in
            if result.last.map({ abs($0 - value) >= TreeRoutingGeometry.epsilon }) ?? true {
                result.append(value)
            }
        }
    }

    private static func terminalContact(
        _ point: CGPoint,
        obstacle: TreeRoutingGeometry.Obstacle
    ) -> Bool {
        let rect = obstacle.rect
        if obstacle.kind == .avatar {
            return ((abs(point.x - rect.minX) < TreeRoutingGeometry.epsilon ||
                abs(point.x - rect.maxX) < TreeRoutingGeometry.epsilon) &&
                point.y >= rect.minY - TreeRoutingGeometry.epsilon &&
                point.y <= rect.maxY + TreeRoutingGeometry.epsilon) ||
                ((abs(point.y - rect.minY) < TreeRoutingGeometry.epsilon ||
                abs(point.y - rect.maxY) < TreeRoutingGeometry.epsilon) &&
                point.x >= rect.minX - TreeRoutingGeometry.epsilon &&
                point.x <= rect.maxX + TreeRoutingGeometry.epsilon)
        }
        return obstacle.kind == .nodeLabel &&
            abs(point.y - rect.maxY - 2) < TreeRoutingGeometry.epsilon &&
            point.x >= rect.minX - TreeRoutingGeometry.epsilon &&
            point.x <= rect.maxX + TreeRoutingGeometry.epsilon
    }

    private static func endpointIsBlocked(
        _ point: CGPoint,
        obstacles: [TreeRoutingGeometry.Obstacle],
        endpointIDs: Set<String>
    ) -> Bool {
        obstacles.contains { obstacle in
            let rect = TreeRoutingGeometry.expanded(
                obstacle.rect,
                by: TreeRoutingGeometry.clearance
            )
            let inside = point.x > rect.minX + TreeRoutingGeometry.epsilon &&
                point.x < rect.maxX - TreeRoutingGeometry.epsilon &&
                point.y > rect.minY + TreeRoutingGeometry.epsilon &&
                point.y < rect.maxY - TreeRoutingGeometry.epsilon
            guard inside else { return false }
            return !endpointIDs.contains(obstacle.ownerID) ||
                !terminalContact(point, obstacle: obstacle)
        }
    }

    private static func accepted(
        _ route: [TreeConnector.Segment],
        obstacles: [TreeRoutingGeometry.Obstacle],
        endpointIDs: Set<String>,
        occupied: [TreeConnector.Segment]
    ) -> Bool {
        TreeRoutingGeometry.routeIsClear(route, obstacles: obstacles, endpointIDs: endpointIDs) &&
            !TreeRoutingGeometry.hasCollinearOverlap(route, occupied: occupied)
    }

    private static func fastCandidates(
        start: CGPoint,
        end: CGPoint,
        obstacles: [TreeRoutingGeometry.Obstacle],
        occupied: [TreeConnector.Segment]
    ) -> [[TreeConnector.Segment]] {
        let clearance = TreeRoutingGeometry.clearance + coordinatePadding
        if start.y == end.y {
            let values = unique(
                obstacles.flatMap { [$0.rect.minY - clearance, $0.rect.maxY + clearance] } +
                    occupied.filter { $0.orientation == .horizontal }
                    .flatMap { [$0.start.y - 6, $0.start.y + 6] }
            ).sorted { abs($0 - start.y) == abs($1 - start.y) ? $0 < $1 : abs($0 - start.y) < abs($1 - start.y) }
            return values.map { y in
                TreeRoutingGeometry.segments(for: [
                    start, CGPoint(x: start.x, y: y), CGPoint(x: end.x, y: y), end,
                ])
            }
        }
        if start.x == end.x {
            let values = unique(
                obstacles.flatMap { [$0.rect.minX - clearance, $0.rect.maxX + clearance] } +
                    occupied.filter { $0.orientation == .vertical }
                    .flatMap { [$0.start.x - 6, $0.start.x + 6] }
            ).sorted { abs($0 - start.x) == abs($1 - start.x) ? $0 < $1 : abs($0 - start.x) < abs($1 - start.x) }
            return values.map { x in
                TreeRoutingGeometry.segments(for: [
                    start, CGPoint(x: x, y: start.y), CGPoint(x: x, y: end.y), end,
                ])
            }
        }
        return []
    }

    private static func escapeXCoordinates(
        point: CGPoint,
        channelY: CGFloat,
        obstacles: [TreeRoutingGeometry.Obstacle],
        occupied: [TreeConnector.Segment]
    ) -> [CGFloat] {
        let lowerY = min(point.y, channelY)
        let upperY = max(point.y, channelY)
        let blockers = obstacles.filter { obstacle in
            point.x > obstacle.rect.minX - TreeRoutingGeometry.clearance &&
                point.x < obstacle.rect.maxX + TreeRoutingGeometry.clearance &&
                lowerY <= obstacle.rect.maxY && upperY >= obstacle.rect.minY
        }
        var candidates = [point.x]
        for blocker in blockers {
            candidates.append(
                blocker.rect.minX - TreeRoutingGeometry.clearance - coordinatePadding
            )
            candidates.append(
                blocker.rect.maxX + TreeRoutingGeometry.clearance + coordinatePadding
            )
        }
        for segment in occupied where segment.orientation == .vertical {
            candidates += [segment.start.x - 6, segment.start.x + 6]
        }
        let sortedCandidates = unique(candidates).sorted { left, right in
            let leftDistance = abs(left - point.x)
            let rightDistance = abs(right - point.x)
            return leftDistance == rightDistance ? left < right : leftDistance < rightDistance
        }
        var values = Array(sortedCandidates.prefix(9))
        if let minX = obstacles.map(\.rect.minX).min(),
           let maxX = obstacles.map(\.rect.maxX).max() {
            for value in [
                minX - TreeRoutingGeometry.clearance - coordinatePadding,
                maxX + TreeRoutingGeometry.clearance + coordinatePadding,
            ] where !values.contains(where: { abs($0 - value) < TreeRoutingGeometry.epsilon }) {
                values.append(value)
            }
        }
        return values
    }

    private static func fallbackRoute(
        start: CGPoint,
        end: CGPoint,
        obstacles: [TreeRoutingGeometry.Obstacle],
        occupied: [TreeConnector.Segment],
        isAccepted: ([TreeConnector.Segment]) -> Bool
    ) -> [TreeConnector.Segment]? {
        if start.x == end.x || start.y == end.y {
            let direct = TreeRoutingGeometry.segments(for: [start, end])
            if isAccepted(direct) { return direct }
        }
        let minObstacleY = min(obstacles.map(\.rect.minY).min() ?? start.y, start.y, end.y)
        let maxObstacleY = max(obstacles.map(\.rect.maxY).max() ?? end.y, start.y, end.y)
        let midpointY = (start.y + end.y) / 2
        var channelCandidates = [
            min(start.y, end.y) - 40,
            max(start.y, end.y) + 40,
            minObstacleY - TreeRoutingGeometry.clearance - coordinatePadding,
            maxObstacleY + TreeRoutingGeometry.clearance + coordinatePadding,
        ]
        for obstacle in obstacles {
            channelCandidates.append(
                obstacle.rect.minY - TreeRoutingGeometry.clearance - coordinatePadding
            )
            channelCandidates.append(
                obstacle.rect.maxY + TreeRoutingGeometry.clearance + coordinatePadding
            )
        }
        for segment in occupied where segment.orientation == .horizontal {
            channelCandidates += [segment.start.y - 6, segment.start.y + 6]
        }
        let channelYs = unique(channelCandidates).sorted { left, right in
            let leftDistance = abs(left - midpointY)
            let rightDistance = abs(right - midpointY)
            return leftDistance == rightDistance ? left < right : leftDistance < rightDistance
        }
        for y in channelYs {
            for startX in escapeXCoordinates(
                point: start,
                channelY: y,
                obstacles: obstacles,
                occupied: occupied
            ) {
                for endX in escapeXCoordinates(
                    point: end,
                    channelY: y,
                    obstacles: obstacles,
                    occupied: occupied
                ) {
                    let candidate = TreeRoutingGeometry.segments(for: [
                        start,
                        CGPoint(x: startX, y: start.y),
                        CGPoint(x: startX, y: y),
                        CGPoint(x: endX, y: y),
                        CGPoint(x: endX, y: end.y),
                        end,
                    ])
                    if isAccepted(candidate) { return candidate }
                }
            }
        }
        return nil
    }

    static func preferredRoute(
        start: CGPoint,
        end: CGPoint,
        obstacles: [TreeRoutingGeometry.Obstacle],
        endpointIDs: Set<String>,
        occupied: [TreeConnector.Segment] = []
    ) -> [TreeConnector.Segment]? {
        let orderedObstacles = sortedObstacles(obstacles)
        guard !endpointIsBlocked(start, obstacles: orderedObstacles, endpointIDs: endpointIDs),
              !endpointIsBlocked(end, obstacles: orderedObstacles, endpointIDs: endpointIDs) else {
            return nil
        }
        let direct = TreeRoutingGeometry.segments(for: [start, end])
        if accepted(
            direct,
            obstacles: orderedObstacles,
            endpointIDs: endpointIDs,
            occupied: occupied
        ) { return direct }
        if let quick = fastCandidates(
            start: start,
            end: end,
            obstacles: orderedObstacles,
            occupied: occupied
        ).first(where: {
            accepted($0, obstacles: orderedObstacles, endpointIDs: endpointIDs, occupied: occupied)
        }) { return quick }
        return fallbackRoute(
            start: start,
            end: end,
            obstacles: orderedObstacles,
            occupied: occupied
        ) {
            accepted($0, obstacles: orderedObstacles, endpointIDs: endpointIDs, occupied: occupied)
        }
    }

    static func routeBetweenPeople(
        left: CGPoint,
        right: CGPoint,
        endpointIDs: Set<String>,
        obstacles: [TreeRoutingGeometry.Obstacle],
        occupied: [TreeConnector.Segment],
        radius: CGFloat
    ) -> [TreeConnector.Segment]? {
        let candidates: [(penalty: CGFloat, start: CGPoint, end: CGPoint)] = [
            (0, CGPoint(x: left.x + radius, y: left.y), CGPoint(x: right.x - radius, y: right.y)),
            (20, CGPoint(x: left.x + radius, y: left.y - 12), CGPoint(x: right.x - radius, y: right.y - 12)),
            (40, CGPoint(x: left.x + radius, y: left.y + 12), CGPoint(x: right.x - radius, y: right.y + 12)),
            (80, CGPoint(x: left.x, y: left.y - radius), CGPoint(x: right.x, y: right.y - radius)),
            (90, CGPoint(x: left.x - 12, y: left.y - radius), CGPoint(x: right.x - 12, y: right.y - radius)),
            (100, CGPoint(x: left.x + 12, y: left.y - radius), CGPoint(x: right.x + 12, y: right.y - radius)),
            (120, CGPoint(x: left.x - radius, y: left.y), CGPoint(x: right.x + radius, y: right.y)),
            (160, CGPoint(x: left.x, y: left.y + radius), CGPoint(x: right.x, y: right.y + radius)),
        ]
        var best: (segments: [TreeConnector.Segment], cost: CGFloat)?
        for candidate in candidates {
            guard let segments = preferredRoute(
                start: candidate.start,
                end: candidate.end,
                obstacles: obstacles,
                endpointIDs: endpointIDs,
                occupied: occupied
            ) else { continue }
            let cost = segments.reduce(0) { $0 + $1.length } +
                CGFloat(max(segments.count - 1, 0)) * bendPenalty + candidate.penalty
            if best == nil || cost < best!.cost { best = (segments, cost) }
        }
        return best?.segments
    }

    static func placeRelationshipLabel(
        relationshipID: String,
        text: String,
        segments: [TreeConnector.Segment],
        obstacles: [TreeRoutingGeometry.Obstacle],
        occupied: [TreeConnector.Segment]
    ) -> (label: TreeRoutingGeometry.RelationshipLabel, obstacle: TreeRoutingGeometry.Obstacle)? {
        let horizontal = segments.filter { $0.orientation == .horizontal }.sorted { left, right in
            if left.length != right.length { return left.length > right.length }
            if left.start.y != right.start.y { return left.start.y < right.start.y }
            return left.start.x < right.start.x
        }
        let fractions = (1...19).map { CGFloat($0) / 20 }.sorted { left, right in
            let leftDistance = abs(left - 0.5)
            let rightDistance = abs(right - 0.5)
            return leftDistance == rightDistance ? left < right : leftDistance < rightDistance
        }
        for segment in horizontal {
            for fraction in fractions {
                let anchor = CGPoint(
                    x: segment.start.x + (segment.end.x - segment.start.x) * fraction,
                    y: segment.start.y
                )
                for offset: CGFloat in [-14, -22, -40, -58, -76, -94] {
                    let center = CGPoint(x: anchor.x, y: anchor.y + offset)
                    let rect = TreeRoutingGeometry.relationshipLabelRect(text: text, center: center)
                    let clearsObstacles = obstacles.allSatisfy {
                        !TreeRoutingGeometry.rectsIntersect(
                            TreeRoutingGeometry.expanded($0.rect, by: TreeRoutingGeometry.clearance),
                            rect
                        )
                    }
                    let clearsRoutes = occupied.allSatisfy {
                        !TreeRoutingGeometry.segmentIntersectsRect($0, rect, clearance: 2)
                    }
                    if clearsObstacles && clearsRoutes {
                        let label = TreeRoutingGeometry.RelationshipLabel(
                            text: text,
                            center: center,
                            rect: rect
                        )
                        return (
                            label,
                            TreeRoutingGeometry.Obstacle(
                                kind: .relationshipLabel,
                                ownerID: relationshipID,
                                rect: rect
                            )
                        )
                    }
                }
            }
        }
        return nil
    }

    static func splitAtAttachmentPoints(
        _ segments: [TreeConnector.Segment]
    ) -> [TreeConnector.Segment] {
        let endpoints = segments.flatMap { [$0.start, $0.end] }
        return segments.flatMap { segment -> [TreeConnector.Segment] in
            let points = endpoints.filter { TreeRoutingGeometry.point($0, isOn: segment) }
                .reduce(into: [CGPoint]()) { result, point in
                    if !result.contains(where: { TreeRoutingGeometry.pointsEqual($0, point) }) {
                        result.append(point)
                    }
                }.sorted {
                    segment.orientation == .horizontal ? $0.x < $1.x : $0.y < $1.y
                }
            guard points.count > 1 else { return [] }
            return zip(points, points.dropFirst()).compactMap { start, end in
                let candidate = TreeConnector.Segment(start: start, end: end)
                return candidate.orientation == nil ? nil : candidate
            }
        }
    }
}
